File: src/C++/mul_matrix.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-16 19:42:50
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-23 19:20:45
@FilePath: \asm_matrix_benchmark\src\C++\mul_matrix.cpp
@Description: mul matrix c++ code
*************************************************************/
#include "matrix_cpp.h"
#include <stdexcept>

/***********************************************************
@description: mul with new matrix
@return {*}
*************************************************************/
MatrixInt MatrixInt::operator*(const MatrixInt &other) const
{
    if (cols != other.rows)
    {
        throw std::invalid_argument("operator*: inner dimensions must match");
    }

    MatrixInt res(rows, other.cols);
    for (size_t i = 0; i < rows; ++i)
    {
        for (size_t j = 0; j < other.cols; ++j)
        {
            long long sum = 0;
            for (size_t k = 0; k < cols; ++k)
            {
                sum += static_cast<long long>(data[i * cols + k]) * other.data[k * other.cols + j];
            }
            res.data[i * res.cols + j] = static_cast<int>(sum);
        }
    }
    return res;
}

/***********************************************************
@description: *=
@return {*}
*************************************************************/
MatrixInt &MatrixInt::operator*=(const MatrixInt &other)
{
    *this = *this * other;
    return *this;
}

/***********************************************************
@description: mul with const number
@return {*}
*************************************************************/
MatrixInt &MatrixInt::operator*=(int scalar)
{
    for (size_t i = 0; i < rows * cols; ++i)
    {
        data[i] *= scalar;
    }
    return *this;
}

File: src/JavaScript/mul_matrix.js
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 20:10:32
@LastEditors: GitHubCopilot
@LastEditTime: 2026-06-29 20:10:32
@FilePath: \asm_matrix_benchmark\src\JavaScript\mul_matrix.js
@Description: mul matrix JavaScript code
*************************************************************/

"use strict";

const { checkMatrix, buildTypedMatrix } = require("./common");

/**
 * Generic matrix multiplication (row-major).
 * @param {{rows:number, cols:number, data:number[]}} m1
 * @param {{rows:number, cols:number, data:number[]}} m2
 * @param {"int"|"float"|"double"} type
 * @returns {{rows:number, cols:number, data:number[]}|null}
 */
function mulMatrix(m1, m2, type) {
  if (!checkMatrix(m1, "m1") || !checkMatrix(m2, "m2")) {
    return null;
  }
  if (m1.cols !== m2.rows) {
    console.log(
      `Dimension mismatch! m1(${m1.rows}, ${m1.cols}) vs m2(${m2.rows}, ${m2.cols})`,
    );
    return null;
  }

  const rows = m1.rows;
  const cols = m2.cols;
  const inner = m1.cols;
  const data = new Array(rows * cols).fill(0);

  // Triple loop is kept intentionally to stay faithful to the C baseline.
  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      let sum = 0;
      for (let k = 0; k < inner; k++) {
        sum += m1.data[i * m1.cols + k] * m2.data[k * m2.cols + j];
      }
      data[i * cols + j] = sum;
    }
  }

  return buildTypedMatrix(rows, cols, data, type);
}

function mul_matrix_int(m1, m2) {
  return mulMatrix(m1, m2, "int");
}

function mul_matrix_float(m1, m2) {
  return mulMatrix(m1, m2, "float");
}

function mul_matrix_double(m1, m2) {
  return mulMatrix(m1, m2, "double");
}

module.exports = {
  mul_matrix_int,
  mul_matrix_float,
  mul_matrix_double,
};
